File: demos/matlab/demo_completion_MNIST.m
%% Demo TOY COMPLETE MNIST
% Illustration of missing data estimation on the MNIST image dataset.
close 
clear
addpath(genpath('../../src/')); % location of GLFM source code
rng( round(sum(1e5*clock)) );

%% LOAD MNIST dataset
% This DB can be downloaded here: http://yann.lecun.com/exdb/mnist/
images = loadMNISTImages('../../datasets/raw/MNIST/train-images-idx3-ubyte');
N = 50;                                             % subset of images to consider
% Xtrue is the ground truth data (each column is one pixel)
Xtrue = images(:, randperm(size(images,2),N))' + 1; % N*D matrix

% Build Xmiss (observation matrix with missing values)
perc_missing = 0.3; % total percentage of missings
missing_val = -100; % value that encodes missing
mask_missings = rand(size(Xtrue)) < perc_missing;
Xmiss = Xtrue;
Xmiss(mask_missings) = missing_val;

data.X = Xmiss;
data.C = repmat('n',1,size(Xmiss,2)); % data type, here pixel values onsidered as count data
% idxOrd = find(max(data.X,[],1)>1) ; %
% data.C(idxOrd)='o';

%% Define parameter default values for algorithm
params.bias = 0;        % bias feature (active for all obs. to capture mean effects)
params.s2B = 0.5;         % variance noise for feature matrix
params.alpha = 10;      % concentration parameter for the Indian Buffet Process
params.Niter = 50;      % number of iterations to train the GLFM model
params.maxK = 100;      % maximum number of latent features (for memory allocation)
params.missing = -100;  % values for missings


%% INFERENCE
[Xcompl, hidden] = GLFM_complete(data,[],params); % no need to initialize Z
% Alternatively, we can compute the whole MAP solution X_map
% hidden = GLFM_infer(data,[],params); % no need to initialize Z
% X_map = GLFM_computeMAP(data.C, hidden.Z, hidden, params);

%% Visualize random image
idx = randi(N,1);
figure;
subplot(1,3,1); imagesc(reshape(Xtrue(idx,:),sqrt(784),sqrt(784)) );
subplot(1,3,2); imagesc(reshape(data.X(idx,:),sqrt(784),sqrt(784)) );
subplot(1,3,3); imagesc(reshape(Xcompl(idx,:),sqrt(784),sqrt(784)) );

disp('SUCCESSFUL');

File: demos/R/demo_completion_MNIST.R
#demo_completion_MNIST.R
#'@description Demo for the MNIST dataset, data completion
#'@param alpha is the concentration parameter for the IBP
#'@param N is the number of datapoints
#'@param s2x is the noise variance
#'@param Z: NxK matrix of feature patterns
#'@param Niter: number of iterations for the Gibbs sampler
#'@param maxR: maximum number of categories across all dimensions
#'@param transf_dummie is a 0-1 variable that indicates if there are data transformations
#'@param th is the threshold to filter out the variables that are not significant
#'@return Xcompl is the completed matrix with maximum a posteriori estimates where the
#'missing values were

demo_data_exploration_MNIST<-function(){
rm(list=ls())
graphics.off()
#setwd("src/GLFMR")
setwd("~/Documents/Working_papers/FAP_Rpackage/GLFM/src/GLFMR")
require(ggplot2)
source("GLFM_infer.R")
source("GLFM_complete.R")
source("GLFM_computePDF.R")
source("GLFM_plotPatterns.R")
source("GLFM_computeMAP.R")
datos_mninst <- read.csv(file="../../datasets/mnist_train_small_100.csv", header=TRUE,stringsAsFactors = FALSE, sep=",")
n_labels<-datos_mninst[,1]
Xauxi<-as.matrix(datos_mninst[,2:785],ncol=99,nrow= 784, byrow=TRUE)
Xfull<-matrix((Xauxi),nrow=784,ncol=99)
Xfull<-t(Xfull)
Xfull<-Xfull+1
perc_missing <- 0.3
missing_val <- -100
N<-dim(Xfull)[1]
D<-dim(Xfull)[2]
Cfull<-rep('n',D)
rand_mat<-matrix(runif(N*D), ncol=D, nrow=N)
mask_missings <- (rand_mat < perc_missing)
Xmiss <- Xfull
Xmiss[mask_missings] <- missing_val
data_mnist<-list("X"=Xmiss,"C"=Cfull)
param_names<-c("missing","s2B","alpha","Niter","maxK","bias","transf_dummie")
bias <- 0 
s2B <- 0.5      
alpha <- 10   
Niter <- 1   
maxK <- 100  
missing <- -100
transf_dummie<-FALSE
params<-list(missing,s2B,alpha,Niter,maxK,bias,transf_dummie)
names(params)<-param_names
output2<-GLFM_complete(data_mnist,list(c(),params))
data_mnist$C<-output2$data$C
# Visualization of a random image
idxs_rndrow<-sample(1:dim(Xfull)[1],28,replace = TRUE, prob=rep(1/dim(Xfull)[1],dim(Xfull)[1]))
idxs_rndcol<-sample(1:dim(Xfull)[2],28,replace = TRUE, prob=rep(1/dim(Xfull)[2],dim(Xfull)[2]))
pixels<-Xfull[idxs_rndrow,idxs_rndcol]
image(pixels,col = grey(seq(0, 1, length = 256)))
# Example with missing 
pixels<-Xmiss[idxs_rndrow,idxs_rndcol]
dev.new()
image(pixels,col = grey(seq(0, 1, length = 256)))
# Example with the completed matrix
pixels<-output2$X_compl[idxs_rndrow,idxs_rndcol]
dev.new()
image(pixels,col = grey(seq(0, 1, length = 256)))
return(list("Xcompl"=output2$X_compl))
}
